File: widget-web/src/controller.ts
/**
 * Framework-agnostic verification controller: a small state machine that drives a
 * verification from start through polling to a terminal state. It holds no DOM and
 * is fully unit-testable. The `mount` helper in index.ts wires it to the DOM.
 *
 * The controller talks to YOUR backend (via the injected start/status callbacks),
 * not to the Coordinator directly, so your API key never reaches the browser.
 */

export type VerificationState = "idle" | "pending" | "verified" | "expired" | "error";

export interface Instructions {
  number: string;
  code?: string;
  channel: string;
  action: string;
  deepLink: string;
  expiresAt: string;
}

export interface StartResult {
  sessionId: string;
  instructions: Instructions;
}

export interface StatusResult {
  status: string;
  verifiedMsisdn?: string;
}

export interface ControllerOptions {
  /** Begin a verification (proxied through your backend). */
  start: () => Promise<StartResult>;
  /** Fetch the current status for a session (proxied through your backend). */
  status: (sessionId: string) => Promise<StatusResult>;
}

type Listener = (payload?: unknown) => void;

export interface Controller {
  begin(): Promise<void>;
  poll(): Promise<VerificationState>;
  state(): VerificationState;
  sessionId(): string | undefined;
  instructions(): Instructions | undefined;
  verifiedMsisdn(): string | undefined;
  on(event: "instructions" | "verified" | "expired" | "error", cb: Listener): void;
}

export function createController(opts: ControllerOptions): Controller {
  let state: VerificationState = "idle";
  let sessionId: string | undefined;
  let instr: Instructions | undefined;
  let verified: string | undefined;
  const listeners: Record<string, Listener[]> = {};

  const emit = (event: string, payload?: unknown) => {
    (listeners[event] ?? []).forEach((fn) => fn(payload));
  };

  return {
    on(event, cb) {
      (listeners[event] ??= []).push(cb);
    },

    async begin() {
      if (state !== "idle") return;
      try {
        const r = await opts.start();
        sessionId = r.sessionId;
        instr = r.instructions;
        state = "pending";
        emit("instructions", instr);
      } catch (err) {
        state = "error";
        emit("error", err);
      }
    },

    async poll() {
      if (state !== "pending" || !sessionId) return state;
      try {
        const s = await opts.status(sessionId);
        if (s.status === "verified") {
          state = "verified";
          verified = s.verifiedMsisdn;
          emit("verified", verified);
        } else if (s.status === "expired" || s.status === "failed") {
          state = "expired";
          emit("expired");
        }
      } catch (err) {
        // Treat a failed poll as transient: stay pending and surface the error.
        emit("error", err);
      }
      return state;
    },

    state: () => state,
    sessionId: () => sessionId,
    instructions: () => instr,
    verifiedMsisdn: () => verified,
  };
}

File: sdk-client-flutter/lib/src/controller.dart
import 'package:flutter/foundation.dart';

import 'types.dart';

enum VerificationPhase { choose, starting, instr, verified, expired, error }

typedef StartFn = Future<StartResult> Function(Channel? channel);
typedef StatusFn = Future<StatusResult> Function(String sessionId);

/// Drives a verification from start through polling to a terminal state. It holds
/// no widgets and is a [ChangeNotifier], so the UI rebuilds as the phase changes.
/// It talks to YOUR backend via [start] / [status], not the Coordinator directly.
class VerificationController extends ChangeNotifier {
  VerificationController({
    required this.start,
    required this.status,
    this.channels = const <Channel>[],
  }) : phase = channels.length > 1 ? VerificationPhase.choose : VerificationPhase.starting;

  final StartFn start;
  final StatusFn status;
  final List<Channel> channels;

  VerificationPhase phase;
  Instructions? instructions;
  String? verifiedMsisdn;
  String? _sessionId;

  /// The error that caused the last transition to [VerificationPhase.error], if any.
  Object? lastError;

  String? get sessionId => _sessionId;

  Future<void> begin([Channel? channel]) async {
    instructions = null;
    _sessionId = null;
    lastError = null;
    phase = VerificationPhase.starting;
    notifyListeners();
    try {
      final r = await start(channel);
      _sessionId = r.sessionId;
      instructions = r.instructions;
      phase = VerificationPhase.instr;
    } catch (e) {
      lastError = e;
      phase = VerificationPhase.error;
    }
    notifyListeners();
  }

  Future<VerificationPhase> poll() async {
    if (phase != VerificationPhase.instr || _sessionId == null) return phase;
    // Client-side deadline: once the instructions expire, stop polling and surface
    // expiry even if the backend never reports it.
    final exp = instructions != null ? DateTime.tryParse(instructions!.expiresAt) : null;
    if (exp != null && DateTime.now().isAfter(exp)) {
      phase = VerificationPhase.expired;
      notifyListeners();
      return phase;
    }
    try {
      final s = await status(_sessionId!);
      if (s.status == 'verified') {
        verifiedMsisdn = s.verifiedMsisdn;
        phase = VerificationPhase.verified;
        notifyListeners();
      } else if (s.status == 'expired' || s.status == 'failed') {
        phase = VerificationPhase.expired;
        notifyListeners();
      }
    } catch (_) {
      // Transient: stay pending.
    }
    return phase;
  }

  void reset() {
    phase = channels.length > 1 ? VerificationPhase.choose : VerificationPhase.starting;
    instructions = null;
    _sessionId = null;
    verifiedMsisdn = null;
    notifyListeners();
  }
}
